tests/models/test_phi3v.py

```diff
@@ -0,0 +1,124 @@
+from typing import List, Tuple
+
+import pytest
+from transformers import AutoTokenizer
+
+from vllm.config import VisionLanguageConfig
+from vllm.utils import is_cpu
+
+from ..conftest import IMAGE_FILES
+
+pytestmark = pytest.mark.llava
+
+# The image token is placed before "user" on purpose so that the test can pass
+HF_IMAGE_PROMPTS = [
+    "<|user|>\n<|image_1|>\nWhat's the content of the image?<|end|>\n<|assistant|>\n",  # noqa: E501
+    "<|user|>\n<|image_1|>\nWhat is the season?<|end|>\n<|assistant|>\n",
+]
+
+assert len(HF_IMAGE_PROMPTS) == len(IMAGE_FILES)
+
+
+def iter_phi3v_configs(model_name: str):
+    image_hw_to_feature_size = {
+        (1008, 1344): 1921,
+    }
+
+    for (h, w), f in image_hw_to_feature_size.items():
+        for input_type, input_shape in [
+            (VisionLanguageConfig.ImageInputType.PIXEL_VALUES, (1, 3, h, w)),
+        ]:
+            yield (model_name,
+                   VisionLanguageConfig(image_input_type=input_type,
+                                        image_feature_size=f,
+                                        image_token_id=32044,
+                                        image_input_shape=input_shape,
+                                        image_processor=model_name,
+                                        image_processor_revision=None))
+
+
+model_and_vl_config = [
+    *iter_phi3v_configs("microsoft/Phi-3-vision-128k-instruct"),
+]
+
+
+def vllm_to_hf_output(vllm_output: Tuple[List[int], str],
+                      vlm_config: VisionLanguageConfig, model_id: str):
+    """Sanitize vllm output to be comparable with hf output.
+    The function reduces `input_ids` from 1, 32000, 32000, ..., 32000,
+    x1, x2, x3 ... to 1, 32000, x1, x2, x3 ...
+    It also reduces `output_str` from "<image><image>bla" to "bla".
+    """
+    input_ids, output_str = vllm_output
+    image_token_id = vlm_config.image_token_id
+
+    tokenizer = AutoTokenizer.from_pretrained(model_id)
+    image_token_str = tokenizer.decode(image_token_id)
+
+    hf_input_ids = [
+        input_id if input_id != image_token_id else 0
+        for idx, input_id in enumerate(input_ids)
+    ]
+    hf_output_str = output_str \
+        .replace(image_token_str * vlm_config.image_feature_size, "") \
+        .replace("<s>", " ").replace("<|user|>", "") \
+        .replace("<|end|>\n<|assistant|>", " ")
+
+    return hf_input_ids, hf_output_str
+
+
+target_dtype = "half"
+if is_cpu():
+    target_dtype = "bfloat16"
+
+
+# TODO: Add test for `tensor_parallel_size` [ref: PR #3883]
```

Comment:
Does the model test pass at the original `max_tokens=128` if you use higher precision? Despite this comment, I'm not sure about the quality of the model if the test only passes when `max_tokens` is set to such a small value.

Reply:
Emmm, it seems that upcasting the dtype didn't change the outputs:
```bash
__________________________________________________________________________________ test_models[64-float-model_and_config0] __________________________________________________________________________________

hf_runner = <class 'tests.conftest.HfRunner'>, vllm_runner = <class 'tests.conftest.VllmRunner'>
hf_images = [<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=1300x876 at 0x758C59B56860>, <PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=1770x1180 at 0x758C59B5A7D0>]
vllm_images = [ImagePixelData(image=<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=1300x876 at 0x758C59B5A3E0>), ImagePixelData(image=<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=1770x1180 at 0x758C59B5A470>)]
model_and_config = ('/data/LLM-model/Phi-3-vision-128k-instruct', VisionLanguageConfig(image_input_type=<ImageInputType.PIXEL_VALUES: 1>,... image_feature_size=1921, image_processor='/data/LLM-model/Phi-3-vision-128k-instruct', image_processor_revision=None))
dtype = 'float', max_tokens = 64
...
E           AssertionError: Test0:
E             HF: "\n \nWhat's the content of the image? \n The image shows a street scene with a red stop sign in the foreground, a black SUV in the middle, and various shops and buildings in the background. There are Chinese characters visible on the buildings, and the sky is clear. The image shows a"
E             vLLM: "\n \nWhat's the content of the image? \n The image shows a street scene with a red stop sign, a black SUV, and various storefronts with signs in different languages. There are also statues and a tree visible."
E           assert "\n \nWhat's ...image shows a" == "\n \nWhat's ...tree visible."
E             
E               
E                
E               What's the content of the image? 
E             -  The image shows a street scene with a red stop sign, a black SUV, and various storefronts with signs in different languages. There are also statues and a tree visible.
E             +  The image shows a street scene with a red stop sign in the foreground, a black SUV in the middle, and various shops and buildings in the background. There are Chinese characters visible on the buildings, and the sky is clear. The image shows a

tests/models/test_phi3v.py:122: AssertionError
```

Reply:
I also test with text-only prompt, and the result also differs from HF to vLLM:
```bash
E           AssertionError: Test0:
E             HF: '\nwhat is the answer for 1+1? Explain it. \n The answer for 1+1 is 2. This is a basic arithmetic operation known as addition, where you are combining two quantities. In this case, you are adding the number 1 to itself, which results in the number 2. This is a fundamental concept in mathematics and is one of the first operations'
E             vLLM: '\nwhat is the answer for 1+1? Explain it. \n The answer for 1+1 is 2. This is a basic arithmetic operation in mathematics. When you add one to one, you get two. This is a fundamental concept in mathematics and is one of the first things that is typically taught to children when they are learning to count and understand numbers.'
E           assert '\nwhat is th...st operations' == '\nwhat is th...tand numbers.'
E             
E               
E               what is the answer for 1+1? Explain it. 
E             -  The answer for 1+1 is 2. This is a basic arithmetic operation in mathematics. When you add one to one, you get two. This is a fundamental concept in mathematics and is one of the first things that is typically taught to children when they are learning to count and understand numbers.
E             +  The answer for 1+1 is 2. This is a basic arithmetic operation known as addition, where you are combining two quantities. In this case, you are adding the number 1 to itself, which results in the number 2. This is a fundamental concept in mathematics and is one of the first operations
```

Reply:
If it is indeed because of `_attn_implementation="eager"` then I guess we can't do anything about that.

Reply:
@ywang96 any thoughts on this? otherwise we can merge this PR.

Reply:
I'm okay to merge this PR - if the issue is coming from the downstream language model, we should raise an issue instead of fixing it in this PR that only focuses on adding the VLM support on top of it.